Source/WebCore/platform/graphics/SourceBufferPrivate.cpp

```diff
@@ -223,50 +220,34 @@ void SourceBufferPrivate::rewindOperationState()
     }
 }
 
-void SourceBufferPrivate::appendCompleted(bool parsingSucceeded, Function<void()>&& preAppendCompletedTask)
-{
-    DEBUG_LOG(LOGIDENTIFIER);
-
-    rewindOperationState();
-    if (parsingSucceeded)
-        queueOperation(AppendCompletedOperation { m_abortCount, WTFMove(preAppendCompletedTask) });
-    else
-        queueOperation(ErrorOperation { });
-}
-
 void SourceBufferPrivate::processAppendCompletedOperation(AppendCompletedOperation&& operation)
 {
     DEBUG_LOG(LOGIDENTIFIER);
 
-    operation.preTask();
-
     // Resolve the changes in TrackBuffers' buffered ranges
     // into the SourceBuffer's buffered ranges
     auto trackBuffers = trackBuffersRanges();
     if (isAttached())
         m_client->sourceBufferPrivateTrackBuffersChanged(trackBuffers);
 
-    updateBufferedFromTrackBuffers(trackBuffers, [weakSelf = WeakPtr { *this }, this, operation = WTFMove(operation)] () mutable {
+    Vector<Ref<GenericPromise>> promises;
+    promises.append(updateBufferedFromTrackBuffers(trackBuffers));
+    if (m_client) {
+        if (m_groupEndTimestamp > duration()) {
+            // https://w3c.github.io/media-source/#sourcebuffer-coded-frame-processing
+            // 5. If the media segment contains data beyond the current duration, then run the duration change algorithm with new
+            // duration set to the maximum of the current duration and the group end timestamp.
+            promises.append(m_client->sourceBufferPrivateDurationChanged(m_groupEndTimestamp));
+        }
+        m_client->sourceBufferPrivateReportExtraMemoryCost(totalTrackBufferSizeInBytes());
```

Comment:
There's a change of timing, we are now calling this synchronously, while before it could be called asynchronously.
Is this right?

Reply:
Sending the report to how much memory is used can be done at any time. Previously we serialised the operation, no we run them in parallel because Native Promise::all allows to conveniently do so. 
The order of operation here doesn't matter. We only need them to happen before we complete the appendBuffer operation